### PR TITLE
Disallow spaces between `@`, attribute name and `(` in Swift 6

### PR DESCRIPTION
**Pull request by ahoppen**
- Disallow space between `@` and attribute name in Swift 6 mode
  - Not having a space between `@` and the attribute name (or even having them on different lines) is weird and shouldn’t be allowed
- Disallow space between attribute name and '(' in Swift 6 mode
  - This allows us to resolve disambiguaties of whether a parenthesis belong to an argument to the attribute or if it is eg. the start of a tuple.

### CODE REVIEW

**Comment by hborla** [comment]
Oh, does this fix the issue with writing `@MainActor` with an explicit closure parameter list?

```swift
let closure = { @MainActor (a, b) in
  print(a, b)
}
```

**Comment by ahoppen** [comment]
It doesn’t actually fix the issue because there’s no behavior change (for source compatibility reasons). But it will produce a warning in Swift 5 and an error in Swift 6.

```swift
let closure = { @MainActor (a, b) in // expected-error {{extraneous whitespace between attribute name and '('}}
}
```

My thought was that we should stage this as a diagnostic and then we can change the parsing behavior to unambiguously parse `(a, b)` as the closure’s parameters in a year or so.

**Review by hamishknight — APPROVED** [review]

File: test/Parse/attribute_spacing.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// RUN: %target-typecheck-verify-swift -swift-version 6
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply test/Parse/attribute_spacing.swift:1]: Oh, good catch. I’ve changed it to use `expected-warning` and not pass `-swift-version 6`

File: test/Parse/attribute_spacing.swift [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+@ MainActor  // expected-error {{extraneous whitespace between '@' and attribute name}}
+class Foo {
+  func foo(_ x: @ escaping () -> Int) {} // expected-error {{extraneous whitespace between '@' and attribute name}}
+}
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply test/Parse/attribute_spacing.swift]: I don’t think adding a dedicated test case for the warning is worth it. See https://github.com/apple/swift/pull/71238#discussion_r1472193305

[the remark this answers is not in the document]
Comment (hamishknight) [reply test/Parse/attribute_spacing.swift]: > What do you mean by the newline case?

The case where the `@` is on the previous line

[the remark this answers is not in the document]
Comment (ahoppen) [reply test/Parse/attribute_spacing.swift]: Added a test case for 

```swift
let closure = { @MainActor
  (a, b) in
}
```

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: test/Parse/attribute_spacing.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (hamishknight) [reply test/Parse/attribute_spacing.swift]: Sorry I should have been clearer, I meant e.g:

```swift
@
MainActor
```

[the remark this answers is not in the document]
Comment (ahoppen) [reply test/Parse/attribute_spacing.swift]: https://github.com/apple/swift/pull/71322

**Pull request merged**